persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/idempotency/IdempotencyAnalyzer.java

```diff
@@ -0,0 +1,122 @@
+import org.apache.cassandra.cql3.statements.ModificationStatement;
+import org.apache.cassandra.cql3.statements.SchemaAlteringStatement;
+import org.apache.cassandra.cql3.statements.SelectStatement;
+import org.apache.cassandra.cql3.statements.TruncateStatement;
+import org.apache.cassandra.cql3.statements.UseStatement;
+import org.apache.cassandra.db.marshal.CounterColumnType;
+import org.apache.cassandra.db.marshal.ListType;
+
+public class IdempotencyAnalyzer {
+  private static final Set<Function> NON_IDEMPOTENT_FUNCTION;
+
+  static {
+    Set<Function> allFunctions = new HashSet<>();
+    allFunctions.addAll(TimeFcts.all());
+    allFunctions.addAll(UuidFcts.all());
+    NON_IDEMPOTENT_FUNCTION = allFunctions;
+  }
+
+  public static boolean isIdempotent(CQLStatement statement) {
+    // if any of the BatchStatement is non-idempotent, return false
+    if (statement instanceof BatchStatement) {
+      BatchStatement batchStatement = (BatchStatement) statement;
+      for (ModificationStatement s : batchStatement.getStatements()) {
+        boolean isIdempotent = analyzeStatement(s);
+        if (!isIdempotent) {
+          return false;
+        }
+      }
+    }
+
+    return analyzeStatement(statement);
+  }
+
+  private static boolean analyzeStatement(CQLStatement statement) {
+    if (statement instanceof SelectStatement) {
+      return true;
+    }
+
+    // Truncate, schema changes, and USE are non-idempotent.
+    if (statement instanceof TruncateStatement
+        || statement instanceof SchemaAlteringStatement
+        || statement instanceof UseStatement) {
+      return false;
+    }
+
+    if (statement instanceof ModificationStatement) {
+      ModificationStatement modification = (ModificationStatement) statement;
+
+      // check if it is a LWT
+      if (modification.hasIfExistCondition()
+          || modification.hasIfNotExistCondition()
+          || modification.hasConditions()) {
+        return false;
+      }
+
+      // check if it is updating a Counter
+      if (modification.isCounter()) {
+        return false;
+      }
+
+      // check if contains non-idempotent function
+      for (Function f : modification.getFunctions()) {
+        if (NON_IDEMPOTENT_FUNCTION.contains(f)) {
+          return false;
+        }
+      }
+
+      for (Operation operation : modification.allOperations()) {
+        // check if it is prepending/appending a List
+        if (operation instanceof Lists.Prepender || operation instanceof Lists.Appender) {
+          return false;
+        }
+      }
+
+      for (ColumnDefinition c : modification.updatedColumns()) {
+        // check if it is updating a Counter
+        if (c.type instanceof CounterColumnType) {
+          return false;
+        }
+      }
+    }
+
+    // delete
+    if (statement instanceof DeleteStatement) {
+      for (ColumnDefinition c : ((DeleteStatement) statement).updatedColumns()) {
+        // on a List
+        if (c.type instanceof ListType) {
```

Comment:
Should this be `CollectionType` to handle all collections (including map and set)?